### PR TITLE
Enable residual tagging for multiple systems in the same input file

### LINKED ISSUE #23194: Residual tags are not working with multiple nonlinear systems
## Bug Description
We get an error message that the problem requests the assembly of a residual vector whose tag does not exist on a system. 

```
*** ERROR ***
Cannot zero vector tag with name 'AHOJ' in system 'v' because there is no vector associated with that tag

Stack frames: 17
0: 0   libmesh_opt.0.dylib                 0x00000001060252bc libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 996
1: 1   libmoose-opt.0.dylib                0x000000010559b7d4 moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 748
2: 2   libmoose-opt.0.dylib                0x00000001052049f4 void mooseError<char const (&) [35], TagName, char const (&) [14], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [54]>(char const (&) [35], TagName&&, char const (&) [14], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [54]) + 352
3: 3   libmoose-opt.0.dylib                0x00000001051b0440 SystemBase::zeroTaggedVector(unsigned int) + 284
4: 4   libmoose-opt.0.dylib                0x00000001051a1e00 NonlinearSystemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 452
5: 5   libmoose-opt.0.dylib                0x0000000104db25d0 FEProblemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 1012
6: 6   libmoose-opt.0.dylib                0x0000000104db2080 FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 180
7: 7   libmoose-opt.0.dylib                0x0000000104dce430 FEProblemBase::computeResidualSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) + 136
8: 8   libmoose-opt.0.dylib                0x00000001051b9ab8 NonlinearSystem::solve() + 208
9: 9   libmoose-opt.0.dylib                0x0000000104dd73dc FEProblemBase::solve(unsigned int) + 428
10: 10  libmoose_test-opt.0.dylib           0x0000000102c6ecf8 auto SteadySolve2::execute()::$_0::operator()<unsigned int>(unsigned int const&) const + 72
11: 11  libmoose_test-opt.0.dylib           0x0000000102c6a998 SteadySolve2::execute() + 340
12: 12  libmoose-opt.0.dylib                0x00000001055706ec MooseApp::executeExecutioner() + 268
13: 13  libmoose_test-opt.0.dylib           0x0000000102ebb83c MooseTestApp::executeExecutioner() + 224
14: 14  libmoose-opt.0.dylib                0x000000010556c5f4 MooseApp::run() + 1504
15: 15  moose_test-opt                      0x00000001024b3b68 main + 160
16: 16  dyld                                0x0000000193c8fe50 start + 2544
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```

## Steps to Reproduce
Go to `moose/test/tests/misc/multiple-nl-systems/test.i` and add add a vector tag to a kernel in system `u`. System `v` complains that it cannot zero this tag. 

```
[Problem]
  nl_sys_names = 'u v'
  extra_tag_vectors = 'ahoj' # <<< Added line
[]

[Kernels]
  [diff_u]
    type = Diffusion
    variable = u
    extra_vector_tags = 'ahoj' # <<< Added line
  []
  [diff_v]
    type = Diffusion
    variable = v
  []
  [force]
    type = CoupledForce
    variable = v
    v = u
  []
[]
```

## Impact
Prevents the utilization of the tagging system when using multiple nonlinear systems.

### PR DESCRIPTION
**Pull request by grmnptr**
Closes #23194

### CODE REVIEW

**Review by loganharbour — COMMENTED** [review]
Why? I don't like this. Fande and I thought a long time how to best generalize this and you're nuking it all. Duplicating all of the methods is not the answer.

**Comment by grmnptr** [comment]
Well, we have multiple systems in a problem now which require separate residual tags and as far as I can see (I might be wrong here) residual and solution vectors behave differently. Of course I am always open to suggestions. I am still experimenting with these, trying to change it as little as possible. But thanks for providing feedback at the very first commit so I don't have to waste my time.

**Comment by grmnptr** [comment]
I might also lack the details here. Why did you guys generalize it? Do some of the applications need additional tag types?

**Comment by YaqiWang** [comment]
There could be a misunderstanding. I think tag id/name belongs to FE problem but not individual systems. The same tag id can be used to add solution vectors for multiple systems (we had this in aux system and nonlinear system before having multi-system). Typically it is up to executioner or a system to add a vector to a particular system for a tag id.

**Comment by lindsayad** [comment]
The issue is with residual vector tags @YaqiWang, which is a problem that we didn't have until multi-system. If someone asks to retrieve residual vector tags, before `computeResidualTags` for instance, you must be able to discern residual tags from what system. We cannot accumulate into residual tags from another system ... the residuals will be the wrong length, not have the corresponding dof indices, etc.

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -5780,12 +5782,8 @@ FEProblemBase::computeResidual(const NumericVector<Number> & soln,
                                const unsigned int nl_sys_num)
 {
   setCurrentNonlinearSystem(nl_sys_num);
-  const auto & residual_vector_tags = getVectorTags(Moose::VECTOR_TAG_RESIDUAL);
-
-  _fe_vector_tags.clear();
 
-  for (const auto & residual_vector_tag : residual_vector_tags)
-    _fe_vector_tags.insert(residual_vector_tag._id);
+  _fe_vector_tags = getResidualVectorTags(nl_sys_num);
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:5786]: Like here @YaqiWang... we can't just set `_fe_vector_tags` to all the `VECTOR_TAG_RESIDUAL` vector tags

**Comment by grmnptr** [comment]
Thanks @YaqiWang for weighing in on this topic! The reason why I am doing this is that I'd like to be able to assemble a residual with a tag on one system but not on the other. The current implementation assumes that every tag exist on every libmesh system which is not true. The example on this issue:

```
[Problem]
  nl_sys_names = 'u v'
  extra_tag_vectors = 'ahoj' # <<< Added line
[]

[Kernels]
  [diff_u]
    type = Diffusion
    variable = u
    extra_vector_tags = 'ahoj' # <<< Added line
  []
  [diff_v]
    type = Diffusion
    variable = v
  []
  [force]
    type = CoupledForce
    variable = v
    v = u
  []
[]
```

Result in 


```
*** ERROR ***
Cannot zero vector tag with name 'AHOJ' in system 'v' because there is no vector associated with that tag

Stack frames: 17
0: 0   libmesh_opt.0.dylib                 0x00000001060252bc libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 996
1: 1   libmoose-opt.0.dylib                0x000000010559b7d4 moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 748
2: 2   libmoose-opt.0.dylib                0x00000001052049f4 void mooseError<char const (&) [35], TagName, char const (&) [14], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [54]>(char const (&) [35], TagName&&, char const (&) [14], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [54]) + 352
3: 3   libmoose-opt.0.dylib                0x00000001051b0440 SystemBase::zeroTaggedVector(unsigned int) + 284
4: 4   libmoose-opt.0.dylib                0x00000001051a1e00 NonlinearSystemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 452
5: 5   libmoose-opt.0.dylib                0x0000000104db25d0 FEProblemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 1012
6: 6   libmoose-opt.0.dylib                0x0000000104db2080 FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 180
7: 7   libmoose-opt.0.dylib                0x0000000104dce430 FEProblemBase::computeResidualSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) + 136
8: 8   libmoose-opt.0.dylib                0x00000001051b9ab8 NonlinearSystem::solve() + 208
9: 9   libmoose-opt.0.dylib                0x0000000104dd73dc FEProblemBase::solve(unsigned int) + 428
10: 10  libmoose_test-opt.0.dylib           0x0000000102c6ecf8 auto SteadySolve2::execute()::$_0::operator()<unsigned int>(unsigned int const&) const + 72
11: 11  libmoose_test-opt.0.dylib           0x0000000102c6a998 SteadySolve2::execute() + 340
12: 12  libmoose-opt.0.dylib                0x00000001055706ec MooseApp::executeExecutioner() + 268
13: 13  libmoose_test-opt.0.dylib           0x0000000102ebb83c MooseTestApp::executeExecutioner() + 224
14: 14  libmoose-opt.0.dylib                0x000000010556c5f4 MooseApp::run() + 1504
15: 15  moose_test-opt                      0x00000001024b3b68 main + 160
16: 16  dyld                                0x0000000193c8fe50 start + 2544
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```

**Comment by YaqiWang** [comment]
> The current implementation assumes that every tag exist on every libmesh system which is not true. 

`FEProblemBase::computeResidualTags` is requiring to set the current system `_current_nl_sys` and accept `tags`. I assume we just passed in unavailable tags when calling `FEProblemBase::computeResidualTags` thus the fix should be only on the caller side. I checked the code. We might just need to be more careful in 
```
FEProblemBase::computeResidual(const NumericVector<Number> & soln,
                               NumericVector<Number> & residual,
                               const unsigned int nl_sys_num)
```
when populating the tags:
```
  for (const auto & residual_vector_tag : residual_vector_tags)
    _fe_vector_tags.insert(residual_vector_tag._id);

```
by filter out tags by calling `_current_nl_sys->hasVector(residual_vector_tag._id)`.

**Comment by grmnptr** [comment]
Ahhh thanks @YaqiWang , I'll give this a try, looks like the least intrusive approach!

**Comment by lindsayad** [comment]
Very nice suggestion!

**Comment by grmnptr** [comment]
The only issue with this is that the tag `RESIDUAL` is never registered in libmesh, it has `RHS Vector`. I can add a function that associates the two with each other but that is rather dirty.

File: framework/include/problems/FEProblemBase.h [hunk 2]
```diff
@@ -1195,6 +1195,8 @@ class FEProblemBase : public SubProblem, public Restartable
    */
   virtual void computeResidualTags(const std::set<TagID> & tags);
 
+  virtual TagName libmeshAliasName(const VectorTag & tag);
```
Comment (lindsayad) [inline framework/include/problems/FEProblemBase.h:1198]:
```suggestion
  static TagName libmeshAliasName(const VectorTag & tag);
```

File: framework/src/problems/FEProblemBase.C [hunk 3]
```diff
@@ -5785,11 +5785,24 @@ FEProblemBase::computeResidual(const NumericVector<Number> & soln,
   _fe_vector_tags.clear();
 
   for (const auto & residual_vector_tag : residual_vector_tags)
-    _fe_vector_tags.insert(residual_vector_tag._id);
+  {
+    // we filter out tags which are not present in the current nonlinear system
+    if (_nl[nl_sys_num]->hasVector(libmeshAliasName(residual_vector_tag)))
+      _fe_vector_tags.insert(residual_vector_tag._id);
+  }
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:5792]:
```suggestion
    // we filter out tags which are not present in the current nonlinear system
    if (_nl[nl_sys_num]->hasVector(libmeshAliasName(residual_vector_tag)))
      _fe_vector_tags.insert(residual_vector_tag._id);
```

**Review by loganharbour — APPROVED** [review]

**Review by loganharbour — CHANGES_REQUESTED** [review]
Please drop the first two commits

File: framework/src/problems/FEProblemBase.C [hunk 4]
```diff
@@ -5780,12 +5780,19 @@ FEProblemBase::computeResidual(const NumericVector<Number> & soln,
...
+  // We associate the residual tag with the given residual vector to make sure we
+  // don't filter it out below
+  _current_nl_sys->associateVectorToTag(residual, _current_nl_sys->residualVectorTag());
   const auto & residual_vector_tags = getVectorTags(Moose::VECTOR_TAG_RESIDUAL);
 
   _fe_vector_tags.clear();
 
   for (const auto & residual_vector_tag : residual_vector_tags)
-    _fe_vector_tags.insert(residual_vector_tag._id);
+    // We filter out tags which are not present in the current nonlinear system. This is
```
Comment (YaqiWang) [inline framework/src/problems/FEProblemBase.C:5792]: `which are not present` => `which do not have associated solution vectors`. This is more accurate.

**Review by lindsayad — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]
Failure is unrelated

**Pull request merged**